src/leap/common/events/client.py

```diff
@@ -55,275 +54,269 @@ class CallbackAlreadyRegistered(Exception):
+        logger.debug("Subscribing to signal: %s" % signal)
+        if not uid:
+            uid = uuid.uuid4()
+        elif uid in self._callbacks[signal] and not replace:
+            raise CallbackAlreadyRegistered()
+        self._callbacks[signal].append((uid, callback))
+        self._sub.socket.setsockopt(zmq.SUBSCRIBE, signal)
+        return uid
 
-    @classmethod
-    def ensure(cls, port):
+    def unregister(self, signal, uid=None):
         """
-        Make sure the daemon is running on the given port.
+        Unregister callbacks for a signal.
 
-        :param port: the port in which the daemon should listen
-        :type port: int
+        If uid is not None, then only the callback identified by the given uid
+        is removed. Otherwise, all callbacks for the signal are removed.
+
+        :param signal: The signal that triggers the callback.
+        :type signal: str
+        :param uid: The callback uid.
+        :type uid: str
+        """
+        self.ensure_client()
+        if not uid:
+            logger.debug(
+                "Unregistering all callbacks from signal %s." % signal)
+            self._callbacks[signal] = {}
+        else:
+            logger.debug(
+                "Unregistering callback %s from signal %s." % (uid, signal))
+            self._callbacks[signal] = filter(
+                lambda u, _: u == uid,
+                self._callbacks[signal])
+        if not self._callbacks[signal]:
+            del self._callbacks[signal]
+            self._sub.socket.setsockopt(zmq.UNSUBSCRIBE, signal)
+
+    def signal(self, signal, content):
+        """
+        Send a signal.
+
+        :param signal: The signal to be sent.
+        :type signal: str
+        :param content: The content of the signal.
+        :type content: str
+        """
+        self.ensure_client()
+        logger.debug("Sending signal: (%s, %s)" % (signal, content))
+        # add send() as a callback for ioloop so it works between threads
+        self._loop.add_callback(
+            lambda: self._push.send(signal + b'\0' + content))
 
-        :return: a daemon instance
-        :rtype: EventsClientDaemon
+    def run(self):
         """
-        return cls.ensure_service(port, EventsClientService())
+        Run the events client.
+        """
+        logger.debug("Starting ioloop.")
+        self._init_zmq()
+        self._initialized.set()
+        self._loop.start()
+        self._loop.close()
+        logger.debug("Ioloop finished.")
+
+    def ensure_client(self):
+        """
+        Make sure the events client thread is started.
+        """
+        with self._lock:
+            if not self.is_alive():
+                self.start()
+                self._initialized.wait()
+
+    def shutdown(self):
+        """
+        Shutdown the events client thread.
+        """
+        with self._lock:
+            if self.is_alive():
+                self._loop.stop(wait=True)
+
+
+_client = EventsClient.instance()
```

Comment:
> hmm, using module-singleton behavior to create the client once?

to do this, you don't need to instantiate the object on each import. it might suffice to keep a module-level container.